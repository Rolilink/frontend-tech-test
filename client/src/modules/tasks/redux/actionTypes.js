import { PENDING, FULFILLED, REJECTED } from 'redux-promise-middleware';

export const FETCH_TASKS = 'travix/front-end-test/tasks/FETCH_TASKS';
export const FETCH_TASKS_SUCCESS = `${FETCH_TASKS}_${FULFILLED}`;
export const FETCH_TASKS_ERROR = `${FETCH_TASKS}_${REJECTED}`;
export const FETCH_TASKS_LOADING = `${FETCH_TASKS}_${PENDING}`;

export const CREATE_TASK = 'travix/front-end-test/tasks/CREATE_TASK';
export const CREATE_TASK_SUCCESS = `${CREATE_TASK}_${FULFILLED}`;
export const CREATE_TASK_ERROR = `${CREATE_TASK}_${REJECTED}`;
export const CREATE_TASK_LOADING = `${CREATE_TASK}_${PENDING}`;
