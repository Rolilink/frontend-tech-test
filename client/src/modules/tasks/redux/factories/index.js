import allTasks from './allTasks';
import http from './http';
import state from './allTasksState';

export {
  allTasks,
  http,
  state,
};
