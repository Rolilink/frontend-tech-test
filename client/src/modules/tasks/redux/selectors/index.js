import getState from './state';
import allTasks from './allTasks';
import createTask from './createTask';
import getShowTasksProps from './showTasksProps';
import getCreateTaskProps from './createTaskProps';

export {
  getState,
  createTask,
  allTasks,
  getShowTasksProps,
  getCreateTaskProps,
};
