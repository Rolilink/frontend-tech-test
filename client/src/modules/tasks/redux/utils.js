export default (tasks) => {
  const normalizedTasks = {};

  tasks.forEach((task) => {
    normalizedTasks[task.id] = task;
  });

  return normalizedTasks;
};
