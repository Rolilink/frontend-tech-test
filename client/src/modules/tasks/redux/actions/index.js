import { createAction } from 'redux-actions';
import { FETCH_TASKS, CREATE_TASK } from '../actionTypes';
import Tasks from '../../../../services/Tasks';

console.log(Tasks);

const tasksService = new Tasks();

export const fetchTasks = createAction(FETCH_TASKS, () => tasksService.fetchTasks());
export const createTask = createAction(CREATE_TASK, task => tasksService.createTask(task));
