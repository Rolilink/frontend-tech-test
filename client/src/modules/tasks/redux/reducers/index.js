import { combineReducers } from 'redux';

import allTasks from './allTasks';
import createTask from './createTask';

export {
  allTasks,
  createTask,
};

export default combineReducers({
  allTasks,
  createTask,
});
