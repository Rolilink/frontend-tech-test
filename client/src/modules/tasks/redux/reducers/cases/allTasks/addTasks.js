/*
 This case adds one or more tasks through an array into the tasks state of all
 tasks, it is created this way to be reused through fetch, update and create.
*/
import { allTasks } from '../../../factories/';
import { normalizeTasks } from '../../../utils';


const tasks = allTasks.tasks;

export default (state, { payload }) => ({
  ...state,
  ...tasks({ tasks: { ...state.tasks, ...normalizeTasks(payload.tasks) } }),
});
