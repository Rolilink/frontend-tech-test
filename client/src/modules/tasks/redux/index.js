import * as actionTypes from './actionTypes';
import * as factories from './factories';
import * as selectors from './selectors';
import * as utils from './utils';

import { fetchTasks, createTask } from './actions';

import reducerModule from './reducers';

const reducer = {
  name: 'tasks', reducer: reducerModule,
};

export {
  fetchTasks,
  createTask,
  actionTypes,
  factories,
  selectors,
  reducer,
  utils,
};
