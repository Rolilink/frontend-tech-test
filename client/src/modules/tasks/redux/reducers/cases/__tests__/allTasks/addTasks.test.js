import addTasks from '../../allTasks/addTasks';
import { actionFactory } from '../../../../../../../__tests__/testUtils/redux';
import { taskFactory } from '../../../../../../../__tests__/testUtils/tasks';
import { normalizeTasks } from '../../../../utils';

jest.mock('../../../../factories/');
const factoriesModule = require('../../../../factories/');

factoriesModule.allTasks.tasks.mockImplementation(state => state);

describe('Tasks.Redux.Reducers.Cases.allTasks.addTasks', () => {
  it('should add tasks to state when state is empty', () => {
    const state = {};
    const tasks = [taskFactory({ id: 1 }), taskFactory({ id: 2 })];
    const action = actionFactory({ payload: { tasks } });

    const newState = addTasks(state, action);

    expect(newState).toMatchSnapshot();
  });

  it('should add tasks withouth erasing existing tasks in the state when the state is not empty', () => {
    const state = {
      tasks: {
        ...normalizeTasks([taskFactory({ id: 1 }), taskFactory({ id: 2 })]),
      },
    };
    const tasks = [taskFactory({ id: 3 }), taskFactory({ id: 5 })];
    const action = actionFactory({ payload: { tasks } });

    const newState = addTasks(state, action);

    expect(newState).toMatchSnapshot();
  });

  it('should update existing tasks in the state when new tasks id matches existing one', () => {
    const state = {
      tasks: {
        ...normalizeTasks([taskFactory({ id: 1 }), taskFactory({ id: 2 })]),
      },
    };
    const tasks = [taskFactory({ id: 2, title: 'updated' })];
    const action = actionFactory({ payload: { tasks } });

    const newState = addTasks(state, action);

    expect(newState).toMatchSnapshot();
  });
});
