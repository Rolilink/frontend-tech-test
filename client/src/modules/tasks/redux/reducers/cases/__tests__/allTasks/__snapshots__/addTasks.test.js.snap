// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks.Redux.Reducers.Cases.allTasks.addTasks should add tasks to state when state is empty 1`] = `
Object {
  "tasks": Object {
    "1": Object {
      "description": "Description",
      "id": 1,
      "title": "Title",
    },
    "2": Object {
      "description": "Description",
      "id": 2,
      "title": "Title",
    },
  },
}
`;

exports[`Tasks.Redux.Reducers.Cases.allTasks.addTasks should add tasks withouth erasing existing tasks in the state when the state is not empty 1`] = `
Object {
  "tasks": Object {
    "1": Object {
      "description": "Description",
      "id": 1,
      "title": "Title",
    },
    "2": Object {
      "description": "Description",
      "id": 2,
      "title": "Title",
    },
    "3": Object {
      "description": "Description",
      "id": 3,
      "title": "Title",
    },
    "5": Object {
      "description": "Description",
      "id": 5,
      "title": "Title",
    },
  },
}
`;

exports[`Tasks.Redux.Reducers.Cases.allTasks.addTasks should update existing tasks in the state when new tasks id matches existing one 1`] = `
Object {
  "tasks": Object {
    "1": Object {
      "description": "Description",
      "id": 1,
      "title": "Title",
    },
    "2": Object {
      "description": "Description",
      "id": 2,
      "title": "updated",
    },
  },
}
`;
